client/executor/wasmtime/src/runtime.rs

```diff
@@ -37,75 +36,96 @@ use sp_runtime_interface::unpack_ptr_and_len;
 use sp_wasm_interface::{Function, Pointer, Value, WordSize};
 use std::{
 	path::{Path, PathBuf},
-	rc::Rc,
 	sync::{
 		atomic::{AtomicBool, Ordering},
 		Arc,
 	},
 };
-use wasmtime::{AsContext, AsContextMut, Engine, StoreLimits};
+use wasmtime::{Engine, Memory, StoreLimits, Table};
 
 pub(crate) struct StoreData {
 	/// The limits we aply to the store. We need to store it here to return a reference to this
 	/// object when we have the limits enabled.
-	limits: StoreLimits,
+	pub(crate) limits: StoreLimits,
 	/// This will only be set when we call into the runtime.
-	host_state: Option<Rc<HostState>>,
+	pub(crate) host_state: Option<HostState>,
+	pub(crate) memory: Option<Memory>,
+	pub(crate) table: Option<Table>,
 }
 
 impl StoreData {
 	/// Returns a reference to the host state.
-	pub fn host_state(&self) -> Option<&Rc<HostState>> {
+	pub fn host_state(&self) -> Option<&HostState> {
 		self.host_state.as_ref()
 	}
+
+	/// Returns a mutable reference to the host state.
+	pub fn host_state_mut(&mut self) -> Option<&mut HostState> {
+		self.host_state.as_mut()
+	}
+
+	/// Returns a reference to the host memory.
+	pub fn memory(&self) -> Memory {
+		self.memory.expect("memory is always set; qed")
+	}
+
+	/// Returns a reference to the host table.
```

Comment:
Hmm... while it is not *exactly* a reference in the Rust sense, it *is* a reference in the general sense. The object returned here is just an opaque handle (think `struct Newtype(usize)`) which points at which slot the object resides in the `Store` - that is, the usual Rust pattern where a `&T` is replaced by an `usize` so that the borrow checker doesn't complain.

So I think calling it a reference here is somewhat apt and was left as-is deliberately. But if you really want I can of course change it. (:

Reply:
I mean I get this, but this sounds like an implementation detail of `Table` to me. That should for sure not be added here, imo.

Reply:
Well, not exactly, because the object also *acts* like a (weak) reference. Notice there's no `clone()` in sight and yet it's returned by value. (: This is definitely part of `wasmtime`'s public API and not an implementation detail. Looking only at the type name you wouldn't know it though.

But, eh, I mean, I don't *really* care either way, so let's not waste time discussing this any further; I'll change this.